qutebrowser/javascript/webelem.js

```diff
@@ -171,14 +211,70 @@ window._qutebrowser.webelem = (function() {
             }
         }
 
+        // Recurse into frames and add them
+        for (let i = 0; i < subelem_frames.length; i++) {
+            if (iframe_same_domain(subelem_frames[i])) {
+                const frame = subelem_frames[i];
+                const subelems = frame.document.
+                    querySelectorAll(selector);
+                for (let elem_num = 0; elem_num < subelems.length; ++elem_num) {
+                    if (!only_visible ||
+                        is_visible(subelems[elem_num], frame)) {
+                        out.push(serialize_elem(subelems[elem_num], frame));
+                    }
+                }
+            }
+        }
+
         return out;
     };
 
+    // Runs a function in a frame until the result is not null, then return
+    function run_frames(func) {
+        for (let i = 0; i < window.frames.length; ++i) {
+            const frame = window.frames[i];
+            if (iframe_same_domain(frame)) {
+                const result = func(frame);
+                if (result) {
+                    return result;
+                }
+            }
+        }
+        return null;
+    }
+
     funcs.find_id = (id) => {
         const elem = document.getElementById(id);
-        return serialize_elem(elem);
+        if (elem) {
+            return serialize_elem(elem);
+        }
+
+        const serialized_elem = run_frames((frame) => {
+            const element = frame.window.document.getElementById(id);
+            return serialize_elem(element, frame);
+        });
+
+        if (serialized_elem) {
+            return serialized_elem;
+        }
+
+        return null;
     };
 
+    // Check if elem is an iframe, and if so, return the result of func on it.
+    // If no iframes match, return null
+    function call_if_frame(elem, func) {
```

Comment:
This seems like it could be a performance hog, as we're looking through all the frames for every element. At the very least we should probably just call it for `iframe` and `frame` objects (or just exit early here). But looking at [this](https://stackoverflow.com/questions/2509934/how-can-i-get-an-element-from-within-a-frameset-frame-using-javascript) and [this](https://stackoverflow.com/questions/1338224/how-do-i-get-the-window-object-from-the-document-object#1338472) SO question, I think there should be a solution which doesn't require iterating over `window.frames`, doing something like `elem.contentDocument.defaultView` or so?

Reply:
Yup, I think we should be able to do this by checking if:

`elem.contentWindow.frameElement` exists

For a fully expanded example:

`document.activeElement.contentWindow.frameElement`

I wonder if there's a better way of checking if something is a frame other than checking if `frameElement` exists.

Reply:
I meant not for checking, but for actually passing it to `func` below, as you don't have to iterate over `window.frames` then.

As for checking: Why not just the equivalent of `elem.tagName() in ['FRAME', 'IFRAME']`?

Reply:
I was not entirely sure if 'frame' and 'iframe' are the only two valid tags for frames, so I wanted to avoid them, if we're sure that that's the case though, that's probably a cleaner solution :)

Reply:
Hmm, fair point. Probably `embed` and `object` too!